apps/system/js/base_icon.js

```diff
@@ -0,0 +1,199 @@
+    this.publish('created');
+  };
+  BaseIcon.prototype = Object.create(BaseUI.prototype);
+  BaseIcon.prototype.constructor = BaseIcon;
+  BaseIcon.prototype.DEBUG = false;
+  BaseIcon.prototype.camalToDash = function(strings) {
+    var i = 0;
+    var ch = '';
+    while (i <= strings.length) {
+      var character = strings.charAt(i);
+      if (character !== character.toLowerCase()) {
+        if (ch === '') {
+          ch += character.toLowerCase();
+        } else {
+          ch += '-' + character.toLowerCase();
+        }
+      } else {
+        ch += character;
+      }
+      i++;
+    }
+    return ch;
+  };
+  BaseIcon.prototype.amendProperty = function() {
+    if (!this.name) {
+      throw new Error('please specify a name when constructing a new icon');
+    }
+    var pureName = this.name.replace(/Icon$/, '');
+    var dashPureName = this.camalToDash(pureName);
+    this.instanceID = 'statusbar-' + dashPureName;
+    if (!this.CLASS_LIST) {
+      this.CLASS_LIST = 'sb-icon sb-icon-' + dashPureName;
+    }
+    if (!this.l10nId) {
+      this.l10nId = 'statusbar' + pureName;
+    }
+  };
+  BaseIcon.prototype.EVENT_PREFIX = 'icon';
+  BaseIcon.prototype.name = 'BaseIcon';
+  BaseIcon.prototype.containerElement = document.getElementById('statusbar');
+  // Overload me
+  BaseIcon.prototype.instanceID = 'statusbar-base';
+  BaseIcon.prototype.additionalProperties = '';
+  BaseIcon.prototype.role = 'listitem';
+  /**
+   * This flag is normally true because most icons want to update
+   * the status once it's rendered. But for certain special icon
+   * they don't want to be updated on started, set this flag to false.
+   * @type {Boolean}
+   */
+  BaseIcon.prototype.UPDATE_ON_START = true;
+  BaseIcon.prototype._fetchElements = function() {
+    this.element = document.getElementById(this.instanceID);
+  };
+  BaseIcon.prototype.show = function(force) {
+    if (!this.element) {
+      return;
+    }
+    var hidden = this.element.hidden;
+    if (!hidden) {
+      return;
+    }
+    this.element.hidden = false;
+    if (!force) {
+      this.publish('shown');
+    }
+  };
+  BaseIcon.prototype.hide = function() {
+    if (!this.element) {
+      return;
+    }
+    var hidden = this.element.hidden;
+    if (hidden) {
+      return;
+    }
+    this.element.hidden = true;
+    this.publish('hidden');
+  };
+  BaseIcon.prototype.render = function() {
+    if (this.element) {
+      this.element.classList.add('active');
+      return false;
+    }
+    var ele = document.getElementById(this.instanceID);
+    if (ele) {
+      this.element = ele;
+      this.element.classList.add('active');
+      this.onrender && this.onrender();
+      this.UPDATE_ON_START && this.update();
+      this.debug('started early');
+    } else {
+      // XXX: To prevent icon ordering regression,
+      // we are waiting statusbar rendered and fetch the element here.
+      Service.request('Statusbar:getIcon', this.instanceID).then(
```

Comment:
I don't think we have a strong enough justification to expose `Statusbar.getIcon` through `Service`.
We're only using the queuing support of `service.js`, but apart from that the `Statusbar` doesn't really need to expose `getIcon` to the world.

I don't see any elegant solution to solve the DOM sharing between the statusbar and icons, but I'd like to settle on something that doesn't "pollute" `Service`.

My suggestion is to dispatch a `statusbarrendered` event at the end of `StatusBar#renderIcons` and to wait for this event instead of doing the `Service.request('Statusbar:getIcon')`.

We could even make a small private (to the `base_icon.js` IIFE) helper in order to end up with something like:
```
    waitForElement().then(function(element, early) {
      this.element = element;
      this.element.classList.toggle('active', early);
      this.onrender && this.onrender();
      this.UPDATE_ON_START && this.update();
    }.bind(this)).catch(function(error) {
      console.log(error);
    });
```
`waitForElement` would maintain a `_statusbarRendered` piece of state global to `base_icons.js` and
* resolve `(element, true)` if `_statusbarRendered` is `false` but `getElementById` returns something
* otherwise wait for the event (and update `_statusbarRendered`) then resolve with `(element, false)`
* and resolve `(element, false)` right away if `_statusbarRendered` is `true`